packages/teams-js/src/internal/validOrigins.ts

```diff
@@ -0,0 +1,125 @@
+import { validOriginsCdnEndpoint, validOriginsFallback } from './constants';
+import { GlobalVars } from './globalVars';
+import { getLogger } from './telemetry';
+import { inServerSideRenderingEnvironment, isValidHttpsURL } from './utils';
+
+let validOriginsCache: string[] = [];
+const validateOriginLogger = getLogger('validateOrigin');
+
+export async function prefetchOriginsFromCDN(): Promise<void> {
+  await getValidOriginsListFromCDN();
+}
+
+function isValidOriginsCacheEmpty(): boolean {
+  return validOriginsCache.length !== 0;
+}
+
+async function getValidOriginsListFromCDN(): Promise<string[]> {
+  if (isValidOriginsCacheEmpty()) {
+    return validOriginsCache;
+  }
+  if (!inServerSideRenderingEnvironment()) {
+    return fetch(validOriginsCdnEndpoint)
+      .then((response) => {
+        if (!response.ok) {
+          throw new Error('Invalid Response from Fetch Call');
+        }
+        return response.json().then((validOriginsCDN) => {
+          if (validateValidOriginsFromCDN(JSON.stringify(validOriginsCDN))) {
+            validOriginsCache = validOriginsCDN.validOrigins;
+            return validOriginsCache;
+          } else {
+            throw new Error('Valid Origins List Is Invalid');
+          }
+        });
+      })
+      .catch((e) => {
+        validateOriginLogger('validOrigins fetch call to CDN failed with error: %s. Defaulting to fallback list', e);
+        validOriginsCache = validOriginsFallback;
+        return validOriginsCache;
+      });
+  } else {
+    validOriginsCache = validOriginsFallback;
+    return validOriginsFallback;
+  }
+}
+
+function validateValidOriginsFromCDN(validOriginsJSON: string): boolean {
+  const validOriginsCDN = JSON.parse(validOriginsJSON);
+  if (!validOriginsCDN.validOrigins) {
+    return false;
+  }
+  for (let i = 0; i < validOriginsCDN.validOrigins.length; i++) {
```

Comment:
Nothing functionally wrong with how you've written this; for future reference I believe the more canonical JS/TS way would be to use a `for..of` loop, e.g.:
```
for (const potentiallyValidOrigin of validOriginsCDN.validOrigins) {
    try {
      new URL('https://' + potentiallyValidOrigin);
    } catch (_) {
      return false;
    }
}
```

Reply:
I blanked out on this nomenclature, thanks I wanted to use this one originally!